packages/react/src/useTableUtils/UseTableDataGenerator.ts

```diff
@@ -0,0 +1,205 @@
+import type { FieldSelection, GadgetRecord } from "@gadgetinc/api-client-core";
+import type { FieldMetadataFragment } from "../internal/gql/graphql.js";
+import { GadgetFieldType } from "../internal/gql/graphql.js";
+import { filterAutoTableFieldList } from "../metadata.js";
+import { isCustomCellColumn, isRelatedFieldColumn } from "../utils.js";
+import type { TableColumn, TableOptions, TableRow } from "./types.js";
+
+type ColumnsOption = Exclude<TableOptions["columns"], undefined>;
+
+const getColumnApiIdentifier = (column: ColumnsOption[number]) => {
+  if (typeof column === "string") {
+    return column;
+  } else if (isRelatedFieldColumn(column)) {
+    return column.field;
+  } else {
+    return column.name;
+  }
+};
+
+const maybeGetRelatedModelFromRelationshipField = (field: FieldMetadataFragment) => {
+  if (
+    field.configuration.__typename === "GadgetBelongsToConfig" ||
+    field.configuration.__typename === "GadgetHasOneConfig" ||
+    field.configuration.__typename === "GadgetHasManyConfig"
+  ) {
+    return field.configuration.relatedModel;
+  }
+};
+
+const getInnerSelectionForRelationshipField = (
+  fieldApiIdentifier: string,
+  relatedFieldApiIdentifier: string,
+  fieldType: GadgetFieldType
+): FieldSelection => {
+  const innerSelection: FieldSelection = {
+    id: true,
+    [relatedFieldApiIdentifier]: true,
+  };
+
+  if (fieldType === GadgetFieldType.HasOne || fieldType === GadgetFieldType.BelongsTo) {
+    return innerSelection;
+  } else if (fieldType === GadgetFieldType.HasMany) {
+    return {
+      edges: {
+        node: innerSelection,
+      },
+    };
+  }
+
+  throw new Error(`Field '${fieldApiIdentifier}' is not a relationship field`);
+};
+
+export class UseTableDataGenerator {
+  readonly targetColumns: ColumnsOption;
+  readonly fieldMetadataMap: Map<string, FieldMetadataFragment>;
+
+  constructor(
+    fieldMetadataArray: FieldMetadataFragment[],
+    columns: TableOptions["columns"],
+    excludeColumns: TableOptions["excludeColumns"],
+    private readonly defaultSelection: Record<string, any>
+  ) {
+    this.fieldMetadataMap = new Map(fieldMetadataArray.map((field) => [field.apiIdentifier, field]));
+
+    if (columns && excludeColumns) {
+      throw new Error("Cannot use both 'columns' and 'excludeColumns' options at the same time");
+    }
+
+    if (columns) {
+      this.targetColumns = columns;
+    } else if (excludeColumns) {
+      const excludeSet = new Set(excludeColumns);
+      this.targetColumns = filterAutoTableFieldList(fieldMetadataArray)
+        .filter((field) => !excludeSet.has(field.apiIdentifier))
+        .map((field) => field.apiIdentifier);
+    } else {
+      this.targetColumns = filterAutoTableFieldList(fieldMetadataArray).map((field) => field.apiIdentifier);
+    }
+  }
+
+  getTableSelectionMap() {
```

Comment:
In here, there should probably be some error for passing `columns={["passwordField"]}` since this will result in this GQL error, which does not obviously indicate that we can't select password fields as columns

![CleanShot 2024-07-26 at 17 45 50@2x](https://github.com/user-attachments/assets/20961f18-abbb-4aa4-807f-6d8d409bf69f)

Reply:
Same thing with hasManyThrough fields. THose are in the metadata but currently can't be queried and don't have a cell renderer